### ISSUE TITLE
Question about modifiers

### ISSUE THREAD

**Issue by zjturner** [facebook/buck2#448]
I was reading through the [Modifiers RFC](https://github.com/facebook/buck2/blob/main/docs/rfcs/drafts/cfg-modifiers/api.md) again since it looks like it's changed since I last read it.

I'm having some trouble reconciling some things that it says:

In the section about [Target Modifiers](https://github.com/facebook/buck2/blob/main/docs/rfcs/drafts/cfg-modifiers/api.md#target-modifier), it states that you can do this:

```
python_binary(
  name = "bar",
  # ...
  cfg_modifiers = {
    "cfg//os:_": "cfg//os:windows",
    # Target modifiers can also use aliases
    "compiler": "clang",
  },
)
```

Later, in the section about [How configuration transitions differ from modifiers](https://github.com/facebook/buck2/blob/main/docs/rfcs/drafts/cfg-modifiers/api.md#how-configuration-modifiers-differ-from-transitions), it says this:

> The difference between modifier and transition is that a transition can change the configuration of any target in the graph, but a modifier can only change the configuration of a top-level target.

In the first example with the `python_binary`, suppose you also have this:

```
python_binary(name = "foo", deps = ["bar"])
```

`bar` is no longer a "top level target" anymore right?  So what does it mean that "a modifier can only change the configuration of a top-level target"?  Does it mean that there are now two versions of `bar`, so that the cgraph looks like this?

```
┌─────────┐             ┌──────────────────────┐
│   foo   │             │         bar          │
│         │             │  <cfg//os:windows>   │
└─────────┘             └──────────────────────┘
     │
     │
     V
┌─────────┐
│   bar   │
└─────────┘
```

After reading the latest changes to the RFC, I was hoping this would offer me a way to have single-target-only constraints, as discussed in [this issue](https://github.com/facebook/buck2/issues/385#issuecomment-1696579365).  It seems like it fits nicely with the design, except for this one little line in the RFC:

> In other words, if you have target A that depends on target B and you request a build of A, then A's target configuration would be resolved via modifiers **and propagated down to B**, but dep B would not do its own modifier resolution.

If there were a way to selectively disable this "propagated down to B" functionality, it sounds like it would solve the problem.

Can anyone shed some light on all of this?

**Comment by cjhopman** [comment]
Yeah, modifiers is mostly about changing how target platform resolution works. target platform resolution only applies to top-level targets and then the resolved configuration is propagated to dependencies. I don't think it will resolve the single target only constraints request.

**Comment by scottcao** [comment]
Can you add some context on what you need single target only constraints for?  I see [this example](https://github.com/facebook/buck2/issues/385#issuecomment-1696567182) from the previous thread but it's just a toy foobar example. What type of rule/target do you want to use single target only constraint for and why can the constraint not propagate to deps?

**Comment by zjturner** [comment]
It’s strictly an optimization to improve cache utilization, but it’s an optimization with potentially massive impact.

The other posts in that thread explain it well I think, but I’ll try again.

Suppose you have an audio library you ship in 2 different products.  The 2 products have different minspec hardware requirements.  One runs on very old devices without, say, AVX instructions.  The other can assume AVX.  You’re going to want to compile these 2 audio libraries differently.  Ideally you represent this with a constraint.

Obviously, in addition to the audio library, these two products share a lot of other code.  Like networking code, a threading library, a standard library of containers, and whatever else.

none of these other libraries care about whether your audio library is built with AVX support, it’s completely unrelated code.

but if you put this constraint in the platform ConfigurationInfo, every single target will now depend on it, and so the configuration hash of the standard library will change depending on whether you’re building Product 1 or Product 2, which is unnecessary and inefficient.  Not only will you cache miss more on all of these libraries, but you’ll fill up the cache with lots of duplicate entries leaving less room for other things.

so I think this is really bad for cache utilization.

This kind of situation happens all the time.  Many configuration options are just fundamentally private to a target for all manner of targets, and the only way to represent that right now (AFAICT) is with buckconfig.

**Comment by scottcao** [comment]
I see, so the concern is with caching. That's something that can be solved in buck2 core if we implement configuration trimming. The deduplication won't be perfect, but the various schemes we've been discussing should all be able to handle the example you showed such that folly doesn't need to be rebuilt twice. We have plans to implement configuration trimming next half.

**Comment by zjturner** [comment]
Yea, configuration trimming was mentioned in the previous thread by @ndmitchell as well, but I was under the impression that the team eventually decided it was too complicated and impossible to implement.  But it sounds like you have a way of making it work now?

**Comment by scottcao** [comment]
There are simpler ways to trim. For example, trimming on the configured graph is relatively easy to implement. It doesn't handle many complicated cases of duplication, but it does handle the case you have in mind. The reason it hasn't been implemented is that it just hasn't been a priority for us. We haven't noticed any impact from configuration divergence internally build speed wise or capacity wise, but we haven't had good ways of measuring that yet either.
